core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/AbstractSolver.java

```diff
@@ -81,9 +82,10 @@ protected void runPhases(SolverScope<Solution_> solverScope) {
         while (!solverTermination.isSolverTerminated(solverScope) && it.hasNext()) {
             Phase<Solution_> phase = it.next();
             phase.solve(solverScope);
-            // If there is a next phase, it starts from the best solution, which might differ from the working solution.
+            // If current phase was not ruin and there is a next phase, it starts from the best solution,
+            // which might differ from the working solution.
             // If there isn't, no need to planning clone the best solution to the working solution.
-            if (it.hasNext()) {
+            if (it.hasNext() && !(phase instanceof RuinCustomPhase)) {
```

Comment:
this condition is still not enough. It may happen that CH and any other phase should also not set working solution from best one

Reply:
Consider two local search configurations, one after another (LS1 and LS2). When LS1 exits, it is very likely that working solution != best solution. Yet LS2 needs to start from the best solution, as starting from the working solution would be a waste. This is in contrast to CH after ruin, which needs to start from the ruined solution, and that - by definition - is not the best solution. 

Practically, it should be the decision of the phase what solution it starts from. I think that the best solution to the problem is to introduce a new method on the phase, something like `pickStartingSolution(...)`. Each phase impl will override this method to implement its own logic for picking the starting solution.

Most phases will start from the best solution, but some may choose to start from the working solution, and that could even be the default impl of the pick method. But if a CH follows a ruin phase, it has to make a choice of not starting from the best solution, but rather from the working solution, overriding the pick method with its own. 

That said, it will be difficult to figure out if a CH follows a ruin, so maybe the switch should instead be "if working solution not initialized, then use working solution, else use best solution".

Reply:
(edited before I saw the answer)

This is not only about CH. This is about every phase. LS should also start from working solution in this case. Lets say CH after ruin ends with a worse score than best. Then we should still start LS from working solution not the best one

Reply:
Good point, didn't think of that. This is a relatively tough nut to crack.

First, we have the backward compatibility consideration - we can not change what LS starts with, we would break every current user. (Are we sure it currently starts from the best solution?) So there would have to be a check that the ruin phase was included, in which case starting from the working solution would be permitted.

Second, what if ruin follows a ruin (ruin-CH-LS-ruin)? Does it start from the best, or the working? Arguably, it starts from the working. So, only the *first* ruin starts from the best solution? 

This is becoming very complicated. We need to figure out a system that *makes sense* and then make it work. As opposed to figuring out what works, and ignoring if it makes sense. :-)

Reply:
It's starting to sound like ruin should be *a move*. Which means it would have to *include*  its own CH. which opens another can of worms. (But eliminates the need for loop phase.)

Reply:
> This is becoming very complicated. We need to figure out a system that _makes sense_ and then make it work. As opposed to figuring out what works, and ignoring if it makes sense. :-)
> It's starting to sound like ruin should be a move inside LS, not its own phase. Which means it would have to include its own CH, which opens another can of worms. (But eliminates the need for loop phase.)

@Christopher-Chianelli , WDYT? Any opinions?

Reply:
I can also think of 3 other approaches:
1. Starting from after ruin phase start using working solution for every phase instead of the best one - relatively easy to do and should do the work for simple cases. On the other hand when the solver config is more complicated and has couple of LS in a row it will make the LS start from worse solution and probably result in a worse final score. 
2. Revert the caching of preruin solution at phase start by ruin phase and comparing at solver end. Also another revert required to make the CH update best solutions again - also easy and has no downsides of the first approach, but it not elegant way I guess
3. Introducing something like bestGlobalScore and solution (something which indeed is the best score) in solver scope? across all the phases used - ? no idea how much work needed, but doesn't seem to be big

Reply:
I'll have to think about this a bit more. For now, I'm thinking something similar to 3, but it needs more time to brew.

Reply:
> > This is becoming very complicated. We need to figure out a system that _makes sense_ and then make it work. As opposed to figuring out what works, and ignoring if it makes sense. :-)
> > It's starting to sound like ruin should be a move inside LS, not its own phase. Which means it would have to include its own CH, which opens another can of worms. (But eliminates the need for loop phase.)
> 
> @Christopher-Chianelli , WDYT? Any opinions?

I was honestly thinking about something similar yesterday, a "paranoid" move that tries to move to a radically different solution by composing many different move. For such a "paranoid" move, it would be helpful if the move itself can evaluate the score in the middle of the move (so it can decide if it need to compose an additional move or not). Alternatively, it might be nice to have access to the score calculator inside a move selector (which allows the selector to see how "good" a particular move is and modify it selection accordingly).

Reply:
Thank you for the suggestion about the "paranoid" move. I may be a good alternative indeed. It sounds to me more like a "Levy flight" than "Ruin&Recreate" principle, which may also do the job.

However I wouldn't like to do this mostly because I am doing it in the scope of my master thesis, which already has "R&R" in its topic, so it is not that flexible and also because my time is limited. We don't know what obstacles we encounter with this new move approach, while "Ruin phase" is pretty much well known at this point and mostly implemented. I still believe we can overcome these issues with assigning best solution to working solution at every phase start with one of the approaches we discussed above.

My suggestion would be to finish what we started with the ruin phase and maybe create another ticket for this paranoid move inside LS. Then at the end of the day when we have both we can compare them and see which approach is better.